src/resources/resource_storage.rs

```diff
@@ -10,14 +10,124 @@ use thiserror::Error;
 use super::{PermissionMask, Resource, ResourceType};
 
 /// Unified resource storage for both redirects and scriptlets.
-#[derive(Default)]
+///
+/// By default, this uses an in-memory storage implementation, however this can be changed using
+/// a custom [ResourceStorageBackend] if desired.
 pub struct ResourceStorage {
+    #[cfg(not(feature = "single-thread"))]
+    backend: Box<dyn ResourceStorageBackend + Sync + Send>,
+    #[cfg(feature = "single-thread")]
+    backend: Box<dyn ResourceStorageBackend>,
+}
+
+/// Loads an empty `InMemoryResourceStorage` backend.
+impl Default for ResourceStorage {
+    fn default() -> Self {
+        Self {
+            backend: Box::new(InMemoryResourceStorage::default()),
+        }
+    }
+}
+
+impl ResourceStorage {
+    #[cfg(not(feature = "single-thread"))]
+    pub fn from_backend<S: ResourceStorageBackend + 'static + Sync + Send>(backend: S) -> Self {
+        Self {
+            backend: Box::new(backend),
+        }
+    }
+
+    #[cfg(feature = "single-thread")]
+    pub fn from_backend<S: ResourceStorageBackend + 'static>(backend: S) -> Self {
+        Self {
+            backend: Box::new(backend),
+        }
+    }
+
+    /// Constructor using an `InMemoryResourceStorage` as the backend with the given resources.
+    #[cfg(test)]
+    pub fn in_memory_from_resources(resources: impl IntoIterator<Item = Resource>) -> Self {
+        Self::from_backend(InMemoryResourceStorage::from_resources(resources))
+    }
+}
+
+/// Customizable backend for [Resource] storage.
+/// Custom implementations could be used to enable (for example) sharing of resources between
+/// multiple [crate::Engine]s, an on-disk backend, or special caching behavior.
+pub trait ResourceStorageBackend {
+    /// Gets the resource associated with `resource_ident`, respecting aliases if necessary.
+    fn get_resource(&self, resource_ident: &str) -> Option<Resource>;
+}
+
+/// Default implementation of [ResourceStorageBackend] that stores all resources in memory.
+#[derive(Default)]
+pub struct InMemoryResourceStorage {
     /// Stores each resource by its canonical name
     resources: HashMap<String, Resource>,
     /// Stores mappings from aliases to their canonical resource names
     aliases: HashMap<String, String>,
 }
 
+impl ResourceStorageBackend for InMemoryResourceStorage {
+    fn get_resource(&self, resource_ident: &str) -> Option<Resource> {
+        let resource = if let Some(resource) = self.resources.get(resource_ident) {
+            Some(resource)
+        } else if let Some(canonical_name) = self.aliases.get(resource_ident) {
+            self.resources.get(canonical_name)
+        } else {
+            None
+        };
+
+        resource.cloned()
+    }
+}
+
+impl InMemoryResourceStorage {
```

Comment:
That code have major conflicts with `master` with https://github.com/brave/adblock-rust/pull/517.